gulpfile.js/build.js

```diff
@@ -392,10 +463,404 @@ function buildPages(done) {
+              if (err) {
+                return callback(err);
+              }
+
+              file.contents = Buffer.from(result);
+              callback(null, file);
+            });
+          })
+        )
+        .pipe(
+          through.obj(function (file, enc, callback) {
+            let renderedPage = file.contents.toString();
+
+            const $ = Cheerio.load(renderedPage);
+
+            $('.nav-link').each(function () {
+              // eslint-disable-next-line no-invalid-this
+              const $this = $(this);
+              const origURL = $this.attr('href');
+              const updatedURL = origURL.replace(
+                coursesRegex,
+                (match, a, b) => `${a}/${coursesPath}/${level}${b}`
+              );
+
+              $this.attr('href', updatedURL);
+            });
+
+            renderedPage = $.root().html();
+
+            if (level === 'beginner') {
+              // eslint-disable-next-line no-invalid-this
+              this.push(
+                new Vinyl({
+                  path: `${file.path}`,
+                  contents: Buffer.from(renderedPage),
+                })
+              );
+            }
+
+            file.path = file.path.replace(
+              coursesRegex,
+              (match, a, b) => `${a}/${coursesPath}/${level}${b}`
+            );
+            file.contents = Buffer.from(renderedPage);
+            callback(null, file);
+          })
+        )
+        .pipe(gulp.dest((f) => f.base))
+        .on('end', cb);
+    };
+
+    Object.defineProperty(f, 'name', {
+      value: `generateCoursesFor${level}`,
+      writable: false,
+    });
+    return f;
+  });
+
+  await gulp
+    .src(`${project.paths.STATICS_DEST}/files/search-promoted-pages/*json`)
+    .pipe(
+      through.obj((file, encoding, callback) => {
+        const rawJSON = file.contents.toString();
+        const data = JSON.parse(rawJSON);
+
+        for (const page of data.pages) {
+          page.url = new URL(page.url, config.hosts.platform.base).toString();
+        }
+
+        for (const page of data.components) {
+          // addExampleAndPlaygroundLink(page, locale);
+          // cleanupTexts(page);
+          page.url = new URL(page.url, config.hosts.platform.base).toString();
+        }
+
+        file.contents = Buffer.from(
+          JSON.stringify({
+            result: data,
+            initial: true,
+          })
+        );
+
+        callback(null, file);
+      })
+    )
+    .pipe(gulp.dest(() => `${project.paths.PAGES_DEST}/search/highlights/`));
+
+  await gulp
+    .src([
+      project.absolute('pages/static/**/*'),
+      project.absolute('examples/static/**/*'),
+      project.absolute('frontend21/dist/static/**/*'),
+    ])
+    .pipe(gulp.dest(`${project.paths.PAGES_DEST}/static`));
+
+  await gulp
+    .src(`${project.paths.STATICS_DEST}/**/*`)
+    .pipe(gulp.dest(() => `${project.paths.PAGES_DEST}/static`));
+
+  return new Promise((resolve, reject) => {
```

Comment:
Couldn't we also return gulp.series instead of a new promise? ....my gulp is a bit rusty.=

Reply:
No, because we need to `resolve` the outer promise we can't return a series. If it wasn't for [this line](https://github.com/ampproject/amp.dev/blob/b1e69255e480c62ea20015342360fd36305d4226/gulpfile.js/build.js#L845) we could do that, however